chainstate/src/detail/chainstateref/consistency_checker.rs

```diff
@@ -0,0 +1,221 @@
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use chainstate_storage::BlockchainStorageRead;
+use chainstate_types::BlockIndex;
+use common::{
+    chain::{Block, ChainConfig, GenBlock},
+    primitives::{BlockHeight, Id},
+};
+use itertools::{EitherOrBoth, Itertools};
+use logging::log;
+
+use super::calc_min_height_with_allowed_reorg;
+
+// Certain tests check for this panic message.
+const PANIC_MSG: &str = "Inconsistent chainstate";
+
+pub struct ConsistencyChecker<'a, 'b, DbTx> {
+    _db_tx: &'a DbTx,
+    chain_config: &'b ChainConfig,
+    /// Keys (block ids) of the block map.
+    block_map_keys: BTreeSet<Id<Block>>,
+    /// The entire block index map.
+    block_index_map: BTreeMap<Id<Block>, BlockIndex>,
+    /// The entire block-by-height map.
+    block_by_height_map: BTreeMap<BlockHeight, Id<GenBlock>>,
+    /// Best block id from the db.
+    best_block_id: Id<GenBlock>,
+    /// The min_height_with_allowed_reorg from the db.
+    min_height_with_allowed_reorg: BlockHeight,
+}
+
+impl<'a, 'b, DbTx: BlockchainStorageRead> ConsistencyChecker<'a, 'b, DbTx> {
+    pub fn new(
+        db_tx: &'a DbTx,
+        chain_config: &'b ChainConfig,
+    ) -> Result<Self, chainstate_storage::Error> {
+        let block_map_keys = db_tx.get_block_map_keys()?;
+        let block_index_map = db_tx.get_block_index_map()?;
+        let block_by_height_map = db_tx.get_block_by_height_map()?;
+        let best_block_id = db_tx.get_best_block_id()?.unwrap_or_else(|| {
+            panic!("{PANIC_MSG}: best block id not stored");
+        });
+        let min_height_with_allowed_reorg =
+            db_tx.get_min_height_with_allowed_reorg()?.unwrap_or(0.into());
+
+        Ok(Self {
+            _db_tx: db_tx,
+            chain_config,
+            block_map_keys,
+            block_index_map,
+            block_by_height_map,
+            best_block_id,
+            min_height_with_allowed_reorg,
+        })
+    }
+
+    pub fn check(&self) {
+        log::debug!("Running chainstate consistency checks");
+
+        self.check_block_index_consistency();
+        self.check_block_height_map_consistency();
+        // TODO: add consistency checks for other maps in the chainstate db.
+    }
+
+    /// Check the block map vs block index map consistency.
+    fn check_block_index_consistency(&self) {
+        for merged in self
```

Comment:
Please add a comment, something in the lines of "Ensure that every block has a corresponding block index, by looping over all blocks and block indices in a merged iterator that indicates whether both are found or only one of either is found"